examples/mem_example/src/main.cpp

```diff
@@ -32,177 +29,197 @@
-                      << message.mean << "; number of data received: "
-                      << ctrImageProcessingServices[4] << std::endl;
-            std::cout << "Event Loop Consumer (" << imageProcessingResults_5
-                      << ") thread ID: " << std::this_thread::get_id()
-                      << std::endl;
-          });
-
-  imageProcessingService_1.startup();
-  imageProcessingService_2.startup();
-  imageProcessingService_3.startup();
-  imageProcessingService_4.startup();
-  imageProcessingService_5.startup();
-  fileImageStreamingService.startup();
-
-  ServerUserInput serverUserInput;
-
-  std::thread imageProcessingServiceThread = std::thread([&]() {
-    while (true) {
-      std::cout << "Image Processing Service Thread thread ID: "
-                << std::this_thread::get_id() << std::endl;
-      fileImageStreamingService.runOnce();
-      imageProcessingService_1.runOnce();
-      imageProcessingService_2.runOnce();
-      imageProcessingService_3.runOnce();
-      imageProcessingService_4.runOnce();
-      imageProcessingService_5.runOnce();
-      std::this_thread::sleep_for(std::chrono::seconds(1));
-      if (!serverUserInput._isRunning) {
-        break;
-      }
+            fileImageStreamingService.runOnce();
+            imageProcessingService_1.runOnce();
+            imageProcessingService_2.runOnce();
+            imageProcessingService_3.runOnce();
+            imageProcessingService_4.runOnce();
+            imageProcessingService_5.runOnce();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            if (!serverUserInput._isRunning) {
+                break;
+            }
+        }
+    });
+
+    serverUserInput.run();
+
+    imageProcessingServiceThread.join();
+
+    fileImageStreamingService.shutdown();
+    imageProcessingService_5.shutdown();
+    imageProcessingService_4.shutdown();
+    imageProcessingService_3.shutdown();
+    imageProcessingService_2.shutdown();
+    imageProcessingService_1.shutdown();
+
+    eventloop.getSubscriber<kpsr_tutorial::ResultData>(imageProcessingResults_5)
+        ->removeListener(imageProcessingResults_5);
+    eventloop.getSubscriber<kpsr_tutorial::ResultData>(imageProcessingResults_4)
+        ->removeListener(imageProcessingResults_4);
+    eventloop.getSubscriber<kpsr_tutorial::ResultData>(imageProcessingResults_3)
+        ->removeListener(imageProcessingResults_3);
+    eventloop.getSubscriber<kpsr_tutorial::ResultData>(imageProcessingResults_2)
+        ->removeListener(imageProcessingResults_2);
+    eventloop.getSubscriber<kpsr_tutorial::ResultData>(imageProcessingResults_1)
+        ->removeListener(imageProcessingResults_1);
+
+    eventloop.stop();
+
+    for (int i = 0; i < nImageProcessingServices; i++) {
+        std::cout << "Total number of data received in imageProcessingService_" << i + 1 << ": "
+                  << ctrImageProcessingServices[i] << std::endl;
```

Comment:
this is probably because the destructors are being called in the wrong order. Please make a jira for this and we can target this later (using gdb/valgrind/address sanitizer tools) in separate pr

Reply:
Kind of improvement after vectorizing the example: garbage instead of errors:

```text
Total number of data received in imageProcessingService_5: 0
Data Multiplexer Consumer (�a��Up�`��Uocessing_results�) thread ID: 140603801765632
Data Multiplexer Consumer (@a��U,a��Uocessing_results0) thread ID: 140603801765632
Data Multiplexer Consumer (@a��U�a��Uocessing_results`) thread ID: 140603801765632
[2023-09-05 12:44:25.442] [debug] kpsr::high_perf::batch_event_processor. alert_ex std::exception on image_processing_feedtest_subscriber.
[2023-09-05 12:44:25.442] [debug] void tp::Worker<Task, Queue>::threadFunc(size_t) [with Task = tp::FixedFunction<void(), 128>; Queue = tp::BlockingQueue; size_t = long unsigned int]. Finished job with name image_processing_feedtest_subscriber.
```

Reply:
> this is probably because the destructors are being called in the wrong order. Please make a jira for this and we can target this later (using gdb/valgrind/address sanitizer tools) in separate pr

https://klepsydra.atlassian.net/browse/KCE-80